src/v/kafka/server/connection_context.cc

```diff
@@ -208,8 +214,13 @@ ss::future<session_resources> connection_context::throttle_request(
     // distinguish throttling delays from real delays. delays
     // applied to subsequent messages allow backpressure to take
     // affect.
-    auto delay = _proto.quota_mgr().record_tp_and_throttle(
-      hdr.client_id, request_size);
+    quota_manager::throttle_delay delay{};
+    if (hdr.key == fetch_api::key) {
+        delay = _proto.quota_mgr().throttle_fetch_tp(hdr.client_id);
+    } else {
+        delay = _proto.quota_mgr().record_produce_tp_and_throttle(
+          hdr.client_id, request_size);
+    }
```

Comment:
Delaying by max of request/response limit is bad here. When client break request size quota limit, all his next requests will be delayed. So if client has produced more than his quota limit, his fetch requests also will be delayed because of request quota. Since delay is calculated from rate and not from current request, that delay can be pretty big.

> In case there are feasible arguments against that, I think record_produce_tp_and_throttle() should then start counting only Produce traffic.

I have left request qouta for all other requests in order to save previous behaviour as much as possible

Reply:
> When client break request size quota limit, all his next requests will be delayed. So if client has produced more than his quota limit, his fetch requests also will be delayed because of request quota. Since delay is calculated from rate and not from current request, that delay can be pretty big.

Wasn't that the behaviour before this change?

I don't mean we shouldn't touch it, but changing this while appealing to preserve the behaviour for all other requests that would suffer from the same problem is probably a suboptimal thing to do. I mean, requests like DescribeCluster, DeleteGroups, OffsetFetch, etc, should not be throttled  neither because of request traffic quota violation, nor because of the response one.

Apart from the question whether should they be throttled or not, there is another question, whether their traffic should be counted towards the quotas or not.

> Delaying by max of request/response limit is bad here

In this context I agree to that, albeit for a different reason.

Let me try to summarize.

| | before | in this PR |
| --- | --- | --- |
| incoming API traffic counted towards **request** quota | all APIs | all APIs except for Fetch |
| APIs throttled for **request** quota violation | all APIs:footprints: | all APIs:dagger: except for Fetch |
| outgoing API traffic counted towards **response** quota | N/A | Fetch only |
| APIs throttled for **response** quota violation | N/A | Fetch only |

My point is, if we argue that :footprints: is not the right way to do, why should we leave :dagger:? I have two ways of thinking how a better way may look like.

| | before | in this PR | alternative A | alternative B |
| --- | --- | --- | --- | --- |
| incoming API traffic counted towards **request** quota | all APIs | all APIs except for Fetch | all APIs | Produce only |
| APIs throttled for **request** quota violation | all APIs | all APIs except for Fetch | Produce only | Produce only |
| outgoing API traffic counted towards **response** quota | N/A | Fetch only | all APIs | Fetch only |
| APIs throttled for **response** quota violation | N/A | Fetch only | Fetch only | Fetch only |

Let me know what you think of it, I would also appreciate @mmaslankaprv's input too, especially regarding the question where we should preserve the existing behaviour as much as possible vs where it's ok to change it.

Reply:
I looked on how it is done in Kafka, and there are multiple quotas defined, there is one global requests quota which is what we have implemented right now. So in the end we should have at least 3 quotas: 
- requests quota - all APIs 
- produce quota - produce only 
- fetch quota - fetch only

It is important to remember that produce/fetch quotas should be throughput based rather than requests count based quotas. 

So the conclusion is that in this PR we introduced fetch quota (which is separate) but we should add fetch requests to our global requests rate quota.

Reply:
Sounds resonable enough. I still wonder why there are both requests quota and produce quota though.

Reply:
the request one is based on the number of requests per second, while the producer one is based on actual throughput

Reply:
In our case, they both are going to be based on actual throughput. Is that ok?